libparsec/crates/types/src/local_manifest/folder.rs

```diff
@@ -215,238 +218,254 @@ impl LocalFolderManifest {
         self.updated = timestamp;
     }
 
-    /// Clear the local confinement points, and remove from the local children any
-    /// entry that was previously confined.
-    ///
-    /// For example, considering a manifest with:
-    /// - `local_confinement_points`: [1, 3]
-    /// - `children`: `{"a.tmp": 1, "b.txt": 2, "c.tmp": 3, "d.txt": 4}`
-    ///
-    /// Then the resulting manifest would have an empty `local_confinement points` and
-    /// `{"b.txt": 2, "d.txt": 4}` as `children`.
-    fn filter_local_confinement_points(mut self) -> Self {
-        if self.local_confinement_points.is_empty() {
-            return self;
-        }
-
-        self.children
-            .retain(|_, entry_id| !self.local_confinement_points.contains(entry_id));
+    pub fn apply_prevent_sync_pattern(
+        &self,
+        prevent_sync_pattern: &Regex,
+        timestamp: DateTime,
+    ) -> Self {
+        UnconfinedLocalFolderManifest::remove_confinement(self).apply_confinement(
+            self,
+            prevent_sync_pattern,
+            timestamp,
+        )
+    }
 
-        self.local_confinement_points.clear();
-        self
+    pub fn from_remote(remote: FolderManifest, prevent_sync_pattern: &Regex) -> Self {
+        UnconfinedLocalFolderManifest::apply_confinement_from_remote(remote, prevent_sync_pattern)
```

Comment:
For me, this method should not exist in `UnconfinedLocalFolderManifest` since it produce a `LocalFolderManifest`.

I would have expected that we go that route: `RemoteFolder -> UnconfinedLocalFolder -> LocalFolder`

Is it possible to make that explicit ?

Reply:
Ahah I kind of hoped it would go unnoticed ^^

So yes, `UnconfinedLocalFolderManifest::apply_confinement_from_remote` is conceptually similar to 
```rust
UnconfinedLocalFolderManifest::from_remote(remote).apply_confinement(None, prevent_sync_pattern, None)
```
if `apply_confinement` were to take `existing_local_manifest` as an `Option<LocalFolderManifest>` (and same for the timestamp).

So I had to choose between:
- 1. Doing just that, using optional arguments, essentially removing `apply_confinement_from_remote` entirely. I didn't like it because `apply_confinement` is already pretty tricky as it is, and it's not obvious at first glance than it will behave correctly for the simple case where the existing manifest doesn't exist. More specifically, in the case of `apply_confinement_from_remote`, you know that the produced `UnconfinedLocalFolderManifest` has its children equal to its base children, which is not the case in general for `apply_confinement` (e.g when it's used to change the prevent sync pattern). I figured, better be explicit with this specific case, as it seems more readable and allows for optimizations.
- 2. I also could have put this specific logic in `LocalFolderManifest::from_remote`, which is where it belongs. But I wanted the confinement handling implementation to be gathered in a single place, for better readability, and so logic can be compared.
- 3. Which is why I ended up specializing `apply_confinement_from_remote` as a dedicated static method.

So I could be convinced that approach 2 is less confusing, but I'd rather stay away from approach 1.

Reply:
So we could move `apply_confinement_from_remote` to `LocalFolderManifest` without putting the logic into `from_remote`.

Reply:
So you're suggesting approach 2 + renaming `LocalFolderManifest::from_remote` to `LocalFolderManifest::apply_confinement_from_remote`? This does not change the point I made in my comment, I would like to keep `apply_confinement_from_remote` close to `apply_confinement` since the latter is a specialized implementation of the former.